packages/flutter/test/services/text_formatter_test.dart

```diff
@@ -605,4 +605,235 @@ void main() {
   });
+
+  group('provided formatters implement shouldReformat correctly', () {
+    test('length limiting formatter', () {
+      expect(
+        LengthLimitingTextInputFormatter(-1).shouldReformat(LengthLimitingTextInputFormatter(null)),
+        isFalse,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(null).shouldReformat(LengthLimitingTextInputFormatter(-1)),
+        isFalse,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(null).shouldReformat(LengthLimitingTextInputFormatter(null)),
+        isFalse,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(3).shouldReformat(LengthLimitingTextInputFormatter(3)),
+        isFalse,
+      );
+
+      // We're relaxing the length constraint. Not reformatting needed.
+      expect(
+        LengthLimitingTextInputFormatter(-1).shouldReformat(LengthLimitingTextInputFormatter(3)),
+        isFalse,
+      );
+
+      // We're relaxing the length constraint. Not reformatting needed.
+      expect(
+        LengthLimitingTextInputFormatter(4).shouldReformat(LengthLimitingTextInputFormatter(3)),
+        isFalse,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(3).shouldReformat(LengthLimitingTextInputFormatter(4)),
+        isTrue,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(3).shouldReformat(LengthLimitingTextInputFormatter(null)),
+        isTrue,
+      );
+
+      expect(
+        LengthLimitingTextInputFormatter(3).shouldReformat(LengthLimitingTextInputFormatter(-1)),
+        isTrue,
+      );
+    });
+
+    test('FliteringTextInputFormatter', () {
+      expect(
+        FilteringTextInputFormatter('a', allow: true, replacementString: 'b').shouldReformat(
+          FilteringTextInputFormatter('a', allow: true, replacementString: 'b'),
+        ),
+        isFalse,
+      );
+
+      expect(
+        FilteringTextInputFormatter('a', allow: true, replacementString: 'b').shouldReformat(
+          FilteringTextInputFormatter('a', allow: true, replacementString: 'c'),
+        ),
+        isTrue,
+      );
+
+      expect(
+        FilteringTextInputFormatter('a', allow: true, replacementString: 'b').shouldReformat(
+          FilteringTextInputFormatter('a', allow: false, replacementString: 'b'),
+        ),
+        isTrue,
+      );
+
+      expect(
+        FilteringTextInputFormatter('a', allow: true, replacementString: 'b').shouldReformat(
+          FilteringTextInputFormatter('c', allow: true, replacementString: 'b'),
+        ),
+        isTrue,
+      );
+
+      expect(
+        FilteringTextInputFormatter('a', allow: true, replacementString: 'b').shouldReformat(
+          FilteringTextInputFormatter('c', allow: true),
+        ),
+        isTrue,
+      );
+    });
+  });
+
+  group('provided formatters do not further modify a formatted value', () {
+    // Framework-provided TextInputFormatters must be idempotent in order to be
```

Comment:
Another place where the term "idempotent" may be misused, see my two previous comments.

Reply:
Here by "idempotent" I mean the format function is guaranteed to converge to a fixed point in one iteration.

Reply:
Got it, we're on the same page now 👍 .  Sorry for the confusion.  I see that the US phone number formatter is described as not idempotent elsewhere now.